### PR TITLE
Default Options for new Volumes

### PR DESCRIPTION
**Pull request by aravindavk**
- Added default group profiles for each volume
  types(replicate,disperse,distribute)
- Included perf xlators in volfile template and disabled in default
  profiles. So that these xlators can be enabled/disabled if
  required.
- Included "features/shard" xlator in client volfile(Fixes: #954)
- Enabled self heal by default for replicate and disperse volumes.
- On glusterd2 start/restart, group profiles are saved in
  `$workdir/templates/profiles.json` (Ex:
  `/var/lib/glusterd2/templates/profiles.json)
- To modify profile defaults, update respective default profile and
  restart glusterd2.(Note: Already created Volume options will not
  change, only applicable for new volumes)
- All perf xlators are disabled by default till we decide the best
  defaults.
- Disabled set/reset of default option profiles.

Signed-off-by: Aravinda VK <avishwan@redhat.com>

### CODE REVIEW

File: glusterd2/commands/volumes/common.go [hunk 1]
```diff
@@ -380,3 +419,25 @@ func txnGenerateBrickVolfiles(c transaction.TxnCtx) error {
...
+				return true
+			}
+		}
+	case []string:
+		for _, v := range value {
+			if strings.HasPrefix(v, pfx) {
+				return true
+			}
+		}
+	default:
```
Comment (Madhu-1) [inline glusterd2/commands/volumes/common.go]: can we discard  default, as we are returning false at the end

Comment (aravindavk) [reply glusterd2/commands/volumes/common.go]: ok

File: glusterd2/commands/volumes/volume-option.go [hunk 2]
```diff
@@ -206,6 +206,11 @@ func volumeOptionsHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if containsReservedGroupProfile(req.Options) {
```
Comment (Madhu-1) [inline glusterd2/commands/volumes/volume-option.go]: do we need this check during volume create also?

**Comment by aravindavk** [comment]
Looks like I need to wait till tomorrow since glusterfs PR(https://review.gluster.org/#/c/glusterfs/+/21813/) merged today

```
17:45:40 			Error:      	Expected nil, but got: &errors.errorString{s:"option not found: performance/io-cache"}
```

Also I missed one thing during volume reset all: Do not reset default profile options.

File: glusterd2/commands/volumes/volume-option.go (hunk 2 above)
Comment (aravindavk) [reply glusterd2/commands/volumes/volume-option.go]: makes sense. I will add it during create as well

**Comment by aravindavk** [comment]
One more fix required in glusterfs. "write-ahead" is used as identifier for write-behind xlator. Because of this option validation is failing. https://github.com/gluster/glusterfs/blob/master/xlators/performance/write-behind/src/write-behind.c#L3258

**Comment by aravindavk** [comment]
This PR is also dependent on https://review.gluster.org/#/c/glusterfs/+/21843/

File: glusterd2/commands/volumes/common.go (hunk 1 above)
Comment (aravindavk) [reply glusterd2/commands/volumes/common.go]: done

File: glusterd2/commands/volumes/volume-option.go (hunk 2 above)
Comment (aravindavk) [reply glusterd2/commands/volumes/volume-option.go]: done

**Comment by aravindavk** [comment]
retest this please

**Comment by aravindavk** [comment]
retest this please

**Comment by aravindavk** [comment]
@Madhu-1 please review

File: glusterd2/commands/volumes/grouped-options.go [hunk 3]
```diff
@@ -6,27 +6,75 @@ import (
...
-		{Name: "cluster/replicate.shd-wait-qlength", OnValue: "10000"},
-		{Name: "features/shard", OnValue: "on"},
-		{Name: "features/shard.shard-block-size", OnValue: "64MB"},
-		{Name: "user.cifs", OnValue: "off"},
-		{Name: "protocol/server.rpc-auth-allow-insecure", OnValue: "on"}},
-		Description: "Enable this profile for optimal results, in block use case"},
-	"profile.SMB-small-file": {Name: "profile.SMB-small-file",
-		Options: []api.VolumeOption{{Name: "features/upcall.cache-invalidation", OnValue: "on"},
+	"profile.default.replicate": {
+		"profile.default.replicate",
```
Comment (Madhu-1) [inline glusterd2/commands/volumes/grouped-options.go]: we may get  `go vet` issues here, please check 
 please check https://github.com/gluster/glusterd2/pull/1377

Comment (aravindavk) [reply glusterd2/commands/volumes/grouped-options.go]: Sorry.. the changes made in your PR got reversed while rebasing. Fixed now

**Comment by Madhu-1** [comment]
@atinmu PTAL

**Comment by Madhu-1** [comment]
@atinmu feel free to merge if this looks good.

**Pull request merged**